### PR TITLE
sct_deepseg_sc implementation

### LINKED ISSUE #1566: implement the deep learning-based cord segmentation
this is becoming quite urgent as more and more users are struggling with propseg.

### PR DESCRIPTION
**Pull request by charleygros**
Pull-request of the sct_deepseg_sc command-line tool that implements a function for the spinal cord segmentation using deep learning algorithm.

Important points:
- Risk of conflicts with [sct_deepseg_gm PR](https://github.com/neuropoly/spinalcordtoolbox/pull/1564).
- Should wait [sct_deepseg_gm PR](https://github.com/neuropoly/spinalcordtoolbox/pull/1564) to be merged for Keras and TensorFlow dependence issues. 

### Description of the Change

- Added deepseg_sc_models to the sct_download_data;
- Added the deepseg_sc API with Keras model;
- Added the sct_deepseg_sc command-line tool (wrapper to the API);
- Added test_sct_deepseg_sc function

### Applicable Issues

Implements or Fixes #1566

### CODE REVIEW

**Comment by jcohenadad** [comment]
i will review after PR #1564 is merged with master (and then master merged with this branch).

**Comment by perone** [comment]
I've just merged the PR #1564 into master.

**Comment by benjamindeleener** [comment]
Thank you @perone 

Master has been merged with this Branch.

@jcohenadad The alternative mirror must be added for `sct_deepseg_sc` models.

File: scripts/sct_download_data.py [hunk 1]
```diff
@@ -92,6 +92,7 @@ def main(args=None):
         'course_hawaii17': 'https://osf.io/6exht/?action=download',
         'deepseg_gm_models': ['https://osf.io/y8k2e/?action=download',
                               'https://www.neuro.polymtl.ca/_media/downloads/sct/20171126_deepgmseg_models.zip'],
+        'deepseg_sc_models': 'https://osf.io/86phg/?action=download'
```
Comment (benjamindeleener) [inline scripts/sct_download_data.py]: The alternative mirror must be provided.

Comment (jcohenadad) [reply scripts/sct_download_data.py]: It's already uploaded (see my email to you and Charley on Jan 25). URL is:
https://www.neuro.polymtl.ca/_media/downloads/sct/20180125_deepseg_sc_models.zip

**Comment by jcohenadad** [comment]
i tried this commit: 7041155ca00c6c7c6a66154e197de1d9952cb69a

i did the following:
~~~
git fetch origin
git checkout -b bdl_scdeepseg origin/bdl_scdeepseg
./install_sct -m -b
~~~

Installation was successful:
~~~
--
Spinal Cord Toolbox (bdl_scdeepseg/7041155ca00c6c7c6a66154e197de1d9952cb69a)
Running /Users/julien/code/sct/scripts/sct_check_dependencies.py
OS: osx (Darwin-14.5.0-x86_64-i386-64bit)
CPU cores: Available: 8, Used by SCT: 8
RAM: Primary memory available: 16.00 gigabytes
SCT path: /Users/julien/code/sct
Installation type: git
  version: dev
  commit: 7041155ca00c6c7c6a66154e197de1d9952cb69a
  branch: bdl_scdeepseg
Check Python executable.............................[OK]
  Using bundled python 2.7.12 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:43:17) 
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)] at /Users/julien/code/sct/python/bin/python
Check if data are installed.........................[OK]
Check if xlwt (1.0.0) is installed..................[OK]
Check if xlutils (1.7.1) is installed...............[OK]
Check if cryptography (1.6) is installed............[OK]
Check if scikit-learn (0.17.1) is installed.........[OK]
Check if scikit-image (0.12.3) is installed.........[OK]
Check if pyqt (4.11.4) is installed.................[OK]
Check if psutil (5.2.2) is installed................[OK]
Check if matplotlib (1.5.1) is installed............[OK]
Check if pip (9.0.1) is installed...................[OK]
Check if requests (2.12.4) is installed.............[OK]
Check if xlrd (0.9.4) is installed..................[OK]
Check if pandas (0.18.1) is installed...............[OK]
Check if mpi4py (2.0.0) is installed................[OK]
Check if dipy (0.11.0) is installed.................[OK]
Check if tqdm (4.11.2) is installed.................[OK]
Check if nibabel (2.1.0) is installed...............[OK]
Check if Keras (2.0.6) is installed.................[OK]
Check if nipy (0.4.1) is installed..................[OK]
Check if tensorflow (1.3.0) is installed............[OK]
Check if h5py (2.7.0) is installed..................[OK]
Check if distribute2mpi (0.3.1) is installed........[OK]
Check if numpy is installed.........................[OK]
Check if scipy is installed.........................[OK]
Check if spinalcordtoolbox is installed.............[OK]
Check ANTs compatibility with OS ...................[OK]
Check PropSeg compatibility with OS ................[OK]
Check if figure can be opened.......................[OK]
~~~

But then I got this error:
~~~
julien-macbook:~ $ sct_deepseg_sc 

--
Spinal Cord Toolbox (bdl_scdeepseg/7041155ca00c6c7c6a66154e197de1d9952cb69a)
Running /Users/julien/code/sct/scripts/sct_deepseg_sc.py
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 29, in <module>
    from spinalcordtoolbox.deepseg_sc.cnn_models import nn_architecture
ImportError: No module named deepseg_sc.cnn_models
~~~

i have then re-installed SCT as follows:
~~~
./install_sct
~~~

and it worked fine. For the future, it would be time-saving to always indicate what are the actions for testing the PR (what to reinstall).
thx

**Comment by jcohenadad** [comment]
@perone: is that normal that `sct_deepgmseg` is still callable? After a full install all aliases should have been erased, no?

**Comment by jcohenadad** [comment]
commit: 7041155ca00c6c7c6a66154e197de1d9952cb69a
data: duke/sct_testing/issues/20180129_deepseg_sc
syntax+error:
~~~
sct_deepseg_sc -i pd_medic.nii.gz -c t2s

--
Spinal Cord Toolbox (bdl_scdeepseg/7041155ca00c6c7c6a66154e197de1d9952cb69a)
Running /Users/julien/code/sct/scripts/sct_deepseg_sc.py -i pd_medic.nii.gz -c t2s
Using TensorFlow backend.
Loading models...
2018-01-29 11:58:51.826720: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2018-01-29 11:58:51.826750: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2018-01-29 11:58:51.826755: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2018-01-29 11:58:51.826758: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
Creating temporary folder...
/var/folders/6f/wy6ljmx9453cgth2qwv5l1s80000gn/T/sct-180129115852-C1sdwF
Reorient the image to RPI, if necessary...
Resample the image to 0.5 mm isotropic resolution...
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 407, in <module>
    main()
  File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 243, in main
    remove_temp_files=remove_temp_files, verbose=verbose, args=args)
  File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 290, in deep_segmentation_spinalcord
    'mm', 'linear', verbose=0)
  File "/Users/julien/code/sct/python/lib/python2.7/site-packages/spinalcordtoolbox/resample/nipy_resample.py", line 153, in resample_file
    interpolation, verbose)
  File "/Users/julien/code/sct/python/lib/python2.7/site-packages/spinalcordtoolbox/resample/nipy_resample.py", line 73, in resample_image
    R[i, i] = n[i] / float(n_r[i])
ZeroDivisionError: float division by zero
~~~

**Comment by jcohenadad** [comment]
@charleygros i got an error because the input file was 4d instead of 3d. It would be good to add a check. Not required for this PR, but please open a ticket.

**Comment by jcohenadad** [comment]
see error here on the DWI scan:
screen shot 2018-01-29 at 14 34 32

**Comment by jcohenadad** [comment]
@benjamindeleener can you please update your review wrt. latest changes? thanks

**Pull request merged**